gsa/src/web/components/form/useForm.js

```diff
@@ -0,0 +1,109 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+import {useState, useEffect} from 'react';
+
+export const capitalizeFirstLetter = string => {
+  if (string.length > 0) {
+    return string[0].toUpperCase() + string.slice(1);
+  }
+  return '';
+};
+
+export const parseAlias = string => {
+  const parsedString = string.replace(/(_)/g, ' ').replace(/([A-Z])/g, ' $1');
+
+  const words = parsedString.split(' ');
+  let errorString = '';
+
+  words.forEach(word => (errorString += capitalizeFirstLetter(word) + ' '));
+  return errorString.trim(); // parse "name" to something human readable. Accepts underscores and camelcase. "credential_id" > "Credential Id", etc.
+};
+
+export const syncVariables = (values, formState, dependencies = {}) => {
+  Object.keys(formState).forEach(key => (values[key] = formState[key]));
+  Object.keys(dependencies).forEach(key => (values[key] = dependencies[key]));
+}; // sync all form states with the values known to SaveDialog
+
+const useForm = (stateSchema, validationRules, callback, extras = {}) => {
+  const [formState, setFormState] = useState(stateSchema);
+  const [dependencies, setDependencies] = useState(extras);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [shouldWarn, setWarn] = useState(false); // shouldWarn is false when first rendered. Only when calling handleSubmit for the first time will this be set to true.
+
+  const [formStatus, setFormStatus] = useState(stateSchema); // use the same shape as stateschema
+
+  const handleValueChange = (value, name) => {
+    setFormState(prevValues => ({...prevValues, [name]: value}));
+  };
+
+  const handleDependencyChange = (value, name) => {
+    setDependencies(prevValues => ({...prevValues, [name]: value}));
+  };
+
+  // eslint-disable-next-line no-shadow
+  const validate = (value, name, dependencies) => {
+    setFormStatus(validity => ({
+      ...validity,
+      [name]: validationRules[name](value, dependencies),
+    }));
+  };
+
+  useEffect(() => {
+    Object.keys(formState).forEach(key => {
+      validate(formState[key], key, dependencies);
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [formState, dependencies]);
+
+  const handleSubmit = vals => {
+    setWarn(true);
+    const errorVars = [];
+    Object.keys(formState).forEach(key => {
+      !formStatus[key].validity && errorVars.push(parseAlias(key)); // Collects form fields with errors
+    });
+
+    const hasErrorInState = Object.keys(formState).some(key => {
+      return formStatus[key].validity === false;
+    }, []); // checks if any field is invalid
+
+    if (hasErrorInState) {
+      setErrorMessage('Error in fields: ' + errorVars.join(', ').toString()); // Outputs an error string containing all errors, or does nothing at all except show error bubbles. This is good for multi step dialogs.
```

Comment:
We need to make sure all strings are translatable.

Reply:
I'm not quite sure how to make dynamically generated error messages translatable. Maybe we need to drop this feature and look for a workaround to achieve the same effect of showing error in fields not rendered.